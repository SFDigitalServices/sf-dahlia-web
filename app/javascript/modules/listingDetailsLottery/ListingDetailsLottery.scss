.modal {
  @apply max-w-0;
}

.modal__inner {
  @apply p-0;
}

.modal__close {
  z-index: 100;
}
