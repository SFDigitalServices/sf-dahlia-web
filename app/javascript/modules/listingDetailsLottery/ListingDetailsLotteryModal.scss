.lottery-modal {
    .lottery-results-form .input {
        border-radius: initial
    }
}