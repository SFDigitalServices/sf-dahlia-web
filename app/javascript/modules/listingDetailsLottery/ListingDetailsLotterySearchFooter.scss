.lottery-results-modal-anchor{
    @apply text-blue-800;
}
