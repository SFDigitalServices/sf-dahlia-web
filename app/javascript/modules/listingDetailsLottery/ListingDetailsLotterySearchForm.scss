.lottery-search-form {
    #search-form input {
        border-radius: initial
    }

    #search-form input::placeholder {
        @apply text-gray-750;
    }
}