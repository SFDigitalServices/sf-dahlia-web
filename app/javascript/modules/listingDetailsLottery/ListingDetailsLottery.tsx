import React, { useEffect, useState } from "react"
import {
  AppearanceSizeType,
  AppearanceStyleType,
  Button,
  Heading,
  Modal,
  t,
} from "@bloom-housing/ui-components"
import { RailsListing } from "../listings/SharedHelpers"
import { isLotteryComplete } from "../../util/listingUtil"
import { getLotteryBucketDetails } from "../../api/listingApiService"
import { RailsLotteryResult } from "../../api/types/rails/listings/RailsLotteryResult"
import { ListingDetailsLotteryModal } from "./ListingDetailsLotteryModal"
import { localizedFormat, renderInlineWithInnerHTML } from "../../util/languageUtil"

export interface ListingDetailsLotteryProps {
  listing: RailsListing
}

export const ListingDetailsLottery = ({ listing }: ListingDetailsLotteryProps) => {
  const [lotteryBucketDetails, setLotteryBucketDetails] = useState<RailsLotteryResult>()
  const [isModalOpen, setIsModalOpen] = useState(false)

  useEffect(() => {
    void getLotteryBucketDetails(listing.Id).then((lotteryBucketDetails) => {
      setLotteryBucketDetails(lotteryBucketDetails)
    })
  }, [listing.Id])

  return (
    isLotteryComplete(listing) && (
      <div className="border-b pt-4 text-center">
        <Heading className="mb-4" priority={4}>
          {t("lottery.lotteryResults")}
        </Heading>
        <p className="mb-4 text-sm uppercase">
          {localizedFormat(listing.Lottery_Results_Date, "LL")}
        </p>
        <div className="bg-gray-100 py-4">
          {listing.Lottery_Summary && (
            <div className="mb-3 mx-2 text-gray-700 text-sm">
              {renderInlineWithInnerHTML(listing.Lottery_Summary)}
            </div>
          )}
          <Button
            size={AppearanceSizeType.small}
            styleType={AppearanceStyleType.primary}
            onClick={() => {
              setIsModalOpen(true)
            }}
          >
            {t("listings.lottery.viewLotteryResults")}
          </Button>
        </div>

        {lotteryBucketDetails && (
          <Modal onClose={() => setIsModalOpen(false)} open={isModalOpen} title="">
            <ListingDetailsLotteryModal
              listing={listing}
              lotteryBucketDetails={lotteryBucketDetails}
            />
          </Modal>
        )}
      </div>
    )
  )
}
