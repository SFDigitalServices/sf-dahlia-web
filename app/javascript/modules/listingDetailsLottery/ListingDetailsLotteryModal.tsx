import React, { useState } from "react"
import { Field, Heading, Icon, IconFillColors, Modal, t } from "@bloom-housing/ui-components"
import { useForm } from "react-hook-form"
import { RailsLotteryBucketsDetails } from "../../api/types/rails/listings/RailsLotteryBucketsDetails"
import { RailsListing } from "../listings/SharedHelpers"
import { ListingDetailsLotteryPreferences } from "./ListingDetailsLotteryPreferences"
import "./ListingDetailsLotteryModal.scss"
import { getLotteryResults } from "../../api/listingApiService"
import { RailsLotteryRanking } from "../../api/types/rails/listings/RailsLotteryRanking"
import { ListingDetailsLotteryRanking } from "./ListingDetailsLotteryRanking"
import { ListingDetailsLotteryModalFooter } from "./ListingDetailsLotteryModalFooter"

export enum STATES {
  INITIAL_STATE,
  LOADING,
  API_ERROR,
  INVALID_LOTTERY_NUMBER,
  SEARCH_RESULT_FOUND,
}

interface ListingDetailsLotteryModalProps {
  isOpen: boolean
  listing: RailsListing
  lotteryBucketDetails: RailsLotteryBucketsDetails
  onClose: () => void
}

export const ListingDetailsLotteryModal = ({
  isOpen,
  listing,
  lotteryBucketDetails,
  onClose,
}: ListingDetailsLotteryModalProps) => {
  // eslint-disable-next-line @typescript-eslint/unbound-method
  const { errors, handleSubmit, register } = useForm({ reValidateMode: "onSubmit" })
  const lotteryNumberField = "lotterySearchNumber"
  const [lotterySearchResult, setLotterySearchResult] = useState<RailsLotteryRanking>()

  const [lotteryModalStatus, setLotteryModalStatus] = useState<STATES>(STATES.INITIAL_STATE)

  const onModalClose = () => {
    setLotteryModalStatus(STATES.INITIAL_STATE)
    onClose()
  }

  const onSubmit = (data: { lotterySearchNumber: string }) => {
    const { lotterySearchNumber } = data
    setLotteryModalStatus(STATES.LOADING)

    void getLotteryResults(listing.Id, lotterySearchNumber).then((lotterySearchResults) => {
      if (!lotterySearchResults) {
        // if we didn't get a result, general search api error
        setLotteryModalStatus(STATES.API_ERROR)
        return
      }

      const { lotteryBuckets } = lotterySearchResults
      if (
        // if we got a result but there are no buckets or none of the buckets have pref
        // results, then we don't have a valid lottery number
        !lotteryBuckets ||
        lotteryBuckets.filter((bucket) => bucket.preferenceResults.length > 0).length === 0
      ) {
        setLotteryModalStatus(STATES.INVALID_LOTTERY_NUMBER)
      } else {
        setLotteryModalStatus(STATES.SEARCH_RESULT_FOUND)
        setLotterySearchResult(lotterySearchResults)
      }
    })
  }

  let content: JSX.Element
  switch (lotteryModalStatus) {
    case STATES.LOADING: {
      content = (
        <div className="text-center">
          <Icon size="large" symbol="spinner" />
        </div>
      )
      break
    }
    case STATES.INVALID_LOTTERY_NUMBER: {
      content = (
        <div className="my-6 text-gray-700">
          <p className="mb-2">{t("lottery.lotteryNumberNotFoundP2")}</p>
          <p className="mb-2">{t("lottery.lotteryNumberNotFoundP3")}</p>
        </div>
      )
      break
    }
    case STATES.SEARCH_RESULT_FOUND: {
      content = lotterySearchResult && (
        <ListingDetailsLotteryRanking lotteryRanking={lotterySearchResult} />
      )
      break
    }
    case STATES.INITIAL_STATE: {
      content = <ListingDetailsLotteryPreferences lotteryBucketsDetails={lotteryBucketDetails} />
      break
    }
  }

  const errorMessage =
    (errors[lotteryNumberField]?.type === "required" && t("lottery.lotteryNumberNotValid")) ||
    (lotteryModalStatus === STATES.INVALID_LOTTERY_NUMBER &&
      t("lottery.lotteryNumberNotFoundP1")) ||
    (lotteryModalStatus === STATES.API_ERROR && t("error.lotteryRankingSearch"))

  return (
    <Modal onClose={onModalClose} open={isOpen} title="">
      <div className="lottery-modal w-96">
        <header className="pb-4 text-center">
          <Heading>{t("lottery.lotteryResults")}</Heading>
          <h2 className="font-sans font-semibold text-sm uppercase">{listing.Name}</h2>
        </header>
        <form
          className="bg-gray-100 flex lottery-results-form mb-4 px-6 py-4"
          onSubmit={handleSubmit(onSubmit)}
        >
          <Field
            error={
              errors[lotteryNumberField] ||
              lotteryModalStatus === STATES.API_ERROR ||
              lotteryModalStatus === STATES.INVALID_LOTTERY_NUMBER
            }
            className="w-full"
            errorMessage={
              errorMessage ||
              (errors[lotteryNumberField]?.type === "required" &&
                t("lottery.lotteryNumberNotValid"))
            }
            register={register}
            name={lotteryNumberField}
            placeholder={t("lottery.enterLotteryNumber")}
            type="text"
            validation={{ required: true }}
          />
          <button className="bg-blue-600 h-12 mt-2 p-3">
            <Icon fill={IconFillColors.white} size="medium" symbol="right" />
          </button>
        </form>
        {content}
        <ListingDetailsLotteryModalFooter
          lotteryModalStatus={lotteryModalStatus}
          listing={listing}
        />
      </div>
    </Modal>
  )
}
