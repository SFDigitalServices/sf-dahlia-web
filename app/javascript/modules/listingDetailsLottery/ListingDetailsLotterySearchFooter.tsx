import React from "react"
import { t } from "@bloom-housing/ui-components"
import { localizedFormat } from "../../util/languageUtil"
import { LOTTERY_SEARCH_FORM_STATUS } from "./ListingDetailsLotterySearchForm"
import { RailsListing } from "../listings/SharedHelpers"

interface ListingDetailsLotterySearchFooterProps {
  lotterySearchFormStatus: LOTTERY_SEARCH_FORM_STATUS
  listing: RailsListing
}

export const ListingDetailsLotterySearchFooter = ({
  lotterySearchFormStatus,
  listing,
}: ListingDetailsLotterySearchFooterProps) => {
  return (
    <>
      <div className="bg-gray-100 mb-2 py-4 text-center">
        <a href="http://sfmohcd.org/housing-preference-programs" target="_blank">
          {t("lottery.readAboutPreferences")}
        </a>
      </div>
      {lotterySearchFormStatus === LOTTERY_SEARCH_FORM_STATUS.INITIAL_STATE && (
        <div className="p-4 text-center text-tiny">
          <p className="py-2">
            {t("lottery.resultsFrom", {
              date: localizedFormat(listing.Lottery_Results_Date, "LL"),
            })}
          </p>
          <p className="py-2">
            <a href={listing.LotteryResultsURL} target="_blank">
              {t("lottery.lotteryResultsPdfLink")}
            </a>
          </p>
          <p className="text-gray-700 py-2">
            {t("lottery.unsortedResultsNote", {
              date: localizedFormat(listing.Lottery_Date, "LL"),
            })}
          </p>
        </div>
      )}
    </>
  )
}
