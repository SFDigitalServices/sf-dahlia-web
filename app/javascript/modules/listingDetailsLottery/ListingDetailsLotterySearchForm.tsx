import React, { useState } from "react"
import { Field, Heading, Icon, IconFillColors, t } from "@bloom-housing/ui-components"
import { useForm } from "react-hook-form"
import type { RailsLotteryResult } from "../../api/types/rails/listings/RailsLotteryResult"
import { RailsListing } from "../listings/SharedHelpers"
import { ListingDetailsLotteryPreferences } from "./ListingDetailsLotteryPreferences"
import { getLotteryResults } from "../../api/listingApiService"
import { ListingDetailsLotteryRanking } from "./ListingDetailsLotteryRanking"
import { ListingDetailsLotterySearchFooter } from "./ListingDetailsLotterySearchFooter"
import "./ListingDetailsLotterySearchForm.scss"
import ErrorBoundary, { BoundaryScope } from "../../components/ErrorBoundary"

export enum LOTTERY_SEARCH_FORM_STATUS {
  INITIAL_STATE,
  LOADING,
  API_ERROR,
  INVALID_LOTTERY_NUMBER,
  SEARCH_RESULT_FOUND,
}

interface ListingDetailsLotterySearchFormProps {
  listing: RailsListing
  lotteryBucketDetails: RailsLotteryResult
}

export const ListingDetailsLotterySearchForm = ({
  listing,
  lotteryBucketDetails,
}: ListingDetailsLotterySearchFormProps) => {
  // eslint-disable-next-line @typescript-eslint/unbound-method
  const { errors, handleSubmit, register } = useForm({ reValidateMode: "onSubmit" })
  const lotteryNumberField = "lotterySearchNumber"
  const [lotterySearchResult, setLotterySearchResult] = useState<RailsLotteryResult>()
  const [lotteryFormStatus, setLotteryFormStatus] = useState<LOTTERY_SEARCH_FORM_STATUS>(
    LOTTERY_SEARCH_FORM_STATUS.INITIAL_STATE
  )

  const onSubmit = (data: { lotterySearchNumber: string }) => {
    const lotterySearchNumber = data.lotterySearchNumber
    setLotteryFormStatus(LOTTERY_SEARCH_FORM_STATUS.LOADING)

    void getLotteryResults(listing.Id, lotterySearchNumber).then((lotterySearchResults) => {
      if (!lotterySearchResults) {
        // if we didn't get a result, general search api error
        setLotteryFormStatus(LOTTERY_SEARCH_FORM_STATUS.API_ERROR)
        return
      }

      const { lotteryBuckets } = lotterySearchResults
      if (
        // if we got a result but there are no buckets or none of the buckets have pref
        // results, then we don't have a valid lottery number
        !lotteryBuckets ||
        lotteryBuckets.filter((bucket) => bucket.preferenceResults.length > 0).length === 0
      ) {
        setLotteryFormStatus(LOTTERY_SEARCH_FORM_STATUS.INVALID_LOTTERY_NUMBER)
      } else {
        setLotteryFormStatus(LOTTERY_SEARCH_FORM_STATUS.SEARCH_RESULT_FOUND)
        setLotterySearchResult(lotterySearchResults)
      }
    })
  }

  let content: JSX.Element
  switch (lotteryFormStatus) {
    case LOTTERY_SEARCH_FORM_STATUS.INITIAL_STATE:
    case LOTTERY_SEARCH_FORM_STATUS.API_ERROR: {
      content = <ListingDetailsLotteryPreferences lotteryBucketsDetails={lotteryBucketDetails} />
      break
    }
    case LOTTERY_SEARCH_FORM_STATUS.LOADING: {
      content = (
        <div className="pb-4 text-center">
          <Icon size="large" symbol="spinner" />
        </div>
      )
      break
    }
    case LOTTERY_SEARCH_FORM_STATUS.INVALID_LOTTERY_NUMBER: {
      content = (
        <div className="my-6 px-8 text-gray-700">
          <p className="mb-2">{t("lottery.lotteryNumberNotFoundP2")}</p>
          <p className="mb-2">{t("lottery.lotteryNumberNotFoundP3")}</p>
        </div>
      )
      break
    }
    case LOTTERY_SEARCH_FORM_STATUS.SEARCH_RESULT_FOUND: {
      content = lotterySearchResult && (
        <ListingDetailsLotteryRanking lotteryResult={lotterySearchResult} />
      )
      break
    }
  }

  const errorMessage =
    (["required", "pattern"].includes(String(errors[lotteryNumberField]?.type)) &&
      t("lottery.lotteryNumberNotValid")) ||
    (lotteryFormStatus === LOTTERY_SEARCH_FORM_STATUS.INVALID_LOTTERY_NUMBER &&
      t("lottery.lotteryNumberNotFoundP1")) ||
    (lotteryFormStatus === LOTTERY_SEARCH_FORM_STATUS.API_ERROR && t("error.lotteryRankingSearch"))

  return (
    <div className="lottery-search-form">
      <header className="pt-8 pb-4 text-center bg-white sticky top-0 z-50">
        <Heading>{t("lottery.lotteryResults")}</Heading>
        <h2 className="font-sans font-semibold text-xs uppercase">{listing.Name}</h2>
      </header>
      <p className="px-8 pb-4 text-xs">{t("lottery.bucketsIntro")}</p>
      <form
        id="search-form"
        className="bg-gray-100 flex mb-4 px-6 py-4"
        // eslint-disable-next-line @typescript-eslint/no-misused-promises
        onSubmit={handleSubmit(onSubmit)}
      >
        <Field
          error={!!errorMessage}
          className="w-full"
          errorMessage={errorMessage}
          register={register}
          name={lotteryNumberField}
          placeholder={t("lottery.enterLotteryNumber")}
          type="text"
          validation={{ pattern: /^\d+$/, required: true }}
          label={t("lottery.enterLotteryNumber")}
          labelClassName="sr-only"
        />
        <button
          className="bg-blue-500 h-12 p-3 translate"
          type="submit"
          aria-label={t("lottery.submitNumber")}
        >
          <Icon fill={IconFillColors.white} size="medium" symbol="right" />
        </button>
      </form>
      <ErrorBoundary boundaryScope={BoundaryScope.component}>
        <div
          aria-live="polite"
          aria-busy={lotteryFormStatus === LOTTERY_SEARCH_FORM_STATUS.LOADING}
        >
          {content}
        </div>
      </ErrorBoundary>
      <ListingDetailsLotterySearchFooter
        lotterySearchFormStatus={lotteryFormStatus}
        listing={listing}
      />
    </div>
  )
}
