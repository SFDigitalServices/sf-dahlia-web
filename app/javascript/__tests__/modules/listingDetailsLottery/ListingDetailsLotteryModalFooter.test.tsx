import React from "react"
import { render } from "@testing-library/react"
import { ListingDetailsLotterySearchFooter } from "../../../modules/listingDetailsLottery/ListingDetailsLotterySearchFooter"
import { lotteryCompleteRentalListing } from "../../data/RailsRentalListing/listing-rental-lottery-complete"
import { LOTTERY_SEARCH_FORM_STATUS } from "../../../modules/listingDetailsLottery/ListingDetailsLotterySearchForm"

describe("ListingDetailsLotteryModalFooter", () => {
  it("displays preference link only when status is not initial state", () => {
    const { asFragment } = render(
      <ListingDetailsLotterySearchFooter
        listing={lotteryCompleteRentalListing}
        lotterySearchFormStatus={LOTTERY_SEARCH_FORM_STATUS.LOADING}
      />
    )

    expect(asFragment()).toMatchSnapshot()
  })
  it("displays full footer when modal status is initial state", () => {
    const { asFragment } = render(
      <ListingDetailsLotterySearchFooter
        listing={lotteryCompleteRentalListing}
        lotterySearchFormStatus={LOTTERY_SEARCH_FORM_STATUS.INITIAL_STATE}
      />
    )

    expect(asFragment()).toMatchSnapshot()
  })
})
