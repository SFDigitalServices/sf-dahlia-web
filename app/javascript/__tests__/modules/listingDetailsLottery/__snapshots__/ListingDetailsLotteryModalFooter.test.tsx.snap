// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListingDetailsLotteryModalFooter displays full footer when modal status is initial state 1`] = `
Array [
  <div
    className="bg-gray-100 py-4 text-center"
  >
    <a
      href="http://sfmohcd.org/housing-preference-programs"
      target="_blank"
    >
      Read About Housing Preferences
    </a>
  </div>,
  <div
    className="p-4 text-center text-tiny"
  >
    <p
      className="py-2"
    >
      Results from January 26, 2022
    </p>
    <p
      className="py-2"
    >
      <a
        href="https://sfmohcd.org/sites/default/files/Documents/MOH/Lottery%20Results/855%20Brannan%20%23318%20Lottery%20Results%201.26.22.pdf"
        target="_blank"
      >
        Download Lottery Results PDF
      </a>
    </p>
    <p
      className="text-gray-700 py-2"
    >
      This PDF contains the unsorted lottery results from the housing lottery held on January 25, 2022.
    </p>
  </div>,
]
`;

exports[`ListingDetailsLotteryModalFooter displays preference link only when status is not initial state 1`] = `
<div
  className="bg-gray-100 py-4 text-center"
>
  <a
    href="http://sfmohcd.org/housing-preference-programs"
    target="_blank"
  >
    Read About Housing Preferences
  </a>
</div>
`;
